Ext auth service : add http/https protocol

norbjd opened the pull request:
<!-- Thanks for sending a pull request! -->

# Changes

:gift: Add http/https protocol for ext auth service

/kind enhancement

Fixes #754

**Release Note**

```release-note
An http/https service can now be used as ext auth service
```

**Docs**

No docs.

norbjd commented:
Hello :wave: 

If this feature is ok to add to Kourier, I will of course write tests to increase code coverage (or at least, to not decrease it :grin:).

nak3 commented:
/ok-to-test

Thank you so much @norbjd 
Let's try to add this fix into v1.4 release as v.1.3 will be released tomorrow (8/Mar).

nak3 commented on pkg/config/ext_authz.go:
```diff
@@ -53,11 +53,32 @@ type ExternalAuthzConfig struct {
+	https extAuthzProtocol = "https"
+)
+
+var extAuthzProtocols = []extAuthzProtocol{grpc, http, https}
```
I think we should use `map`.  Gateway API also uses map like https://github.com/kubernetes-sigs/gateway-api/blob/master/apis/v1alpha2/validation/gateway.go#L33-L38

nak3 commented on pkg/config/ext_authz.go:
```diff
@@ -53,11 +53,32 @@ type ExternalAuthzConfig struct {
+type extAuthzProtocol string
+
+const (
+	grpc  extAuthzProtocol = "grpc"
```
It is better to use `extAuthzProtocolGRPC` or `GRPCextAuthzProtocol` for the name?

nak3 commented on pkg/config/ext_authz.go:
```diff
@@ -88,17 +114,28 @@ func init() {
-		Cluster:    extAuthzCluster(host, uint32(port)),
-		HTTPFilter: externalAuthZFilter(extAuthzClusterName, timeout, env.FailureModeAllow, env.MaxRequestBytes),
+		Cluster:    extAuthzCluster(host, uint32(port), env.Protocol),
+		HTTPFilter: externalAuthZFilter(extAuthzClusterName, env.Host, env.PathPrefix, timeout, env.FailureModeAllow, env.MaxRequestBytes, env.Protocol),
```
I would pass `env` to the args instead instead of five `env.*` variables by changing `externalAuthZFilter()` to accept `*config`.
```suggestion
		HTTPFilter: externalAuthZFilter(extAuthzClusterName, timeout, env),
```

nak3 reviewed (commented):
Some refactoring comments but LGM overall. We would like to have tests as @norbjd already mentioned :+1:

norbjd commented:
/retest

norbjd commented:
Hello @nak3 ! Thanks for your review.

I have refactored the code according to your comments, and also added some unit tests and e2e tests.

Unfortunately, e2e tests (`e2e-kind.sh`) do not pass (I've retried once). The culprit lines are:

https://github.com/knative-sandbox/net-kourier/blob/19369c8b6eb2ed90276c7b2d7c96a7cd59d06bc2/test/e2e-kind.sh#L67-L71

```
>> Setup ExtAuthz gRPC
2022/03/12 15:53:24 Using base gcr.io/distroless/static:nonroot for knative.dev/net-kourier/test/config/extauthz/grpc/src
2022/03/12 15:53:25 Building knative.dev/net-kourier/test/config/extauthz/grpc/src for linux/amd64
[...]
deployment.apps/externalauthz-grpc created
service/externalauthz-grpc created
deployment.apps/externalauthz-grpc condition met
deployment.apps/net-kourier-controller env updated
Waiting for deployment "net-kourier-controller" rollout to finish: 1 old replicas are pending termination...
error: timed out waiting for the condition
Error: Process completed with exit code 1.
```

I don't really understand why the `net-kourier-controller` rollout timeouts after updating the environment variable. I have just created a subfolder `grpc` under `test/config/extauthz`, to add a new subfolder for `http` tests, so I believe the error comes from here.

Would you mind taking a look please :pray: 

Thanks again.

nak3 commented on pkg/config/ext_authz_test.go:
```diff
@@ -0,0 +1,316 @@
+			},
+		},
+		{
+			name: "http_with_path_prefix",
```
```suggestion
			name: "http with path prefix",
```
It is fine to use spalce.

nak3 commented on pkg/config/ext_authz_test.go:
```diff
@@ -0,0 +1,316 @@
+			},
+		},
+		{
+			name: "https_with_path_prefix",
```
```suggestion
			name: "https with path prefix",
```

Same with above.

nak3 commented on pkg/config/ext_authz_test.go:
```diff
@@ -0,0 +1,316 @@
+	for _, tt := range tests {
+		t.Run(string(tt.protocol), func(t *testing.T) {
+			if got := isValidExtAuthzProtocol(tt.protocol); got != tt.want {
+				t.Errorf("isValidExtAuthzProtocol() = %v, extAuthzWanted %v", got, tt.want)
```
```suggestion
				t.Errorf("isValidExtAuthzProtocol() = %v, want = %v", got, tt.want)
```

serving repo uses this style, so let's align with them.

nak3 commented on pkg/config/ext_authz_test.go:
```diff
@@ -0,0 +1,316 @@
+			}
+
+			if !reflect.DeepEqual(httpProtocolOptionsGot, httpProtocolOptionsWantedAny) {
+				t.Errorf("extAuthzCluster() http protocol options = %v, wanted %v", httpProtocolOptionsGot, httpProtocolOptionsWantedAny)
```
```suggestion
				t.Errorf("extAuthzCluster() http protocol options = %v, want = %v", httpProtocolOptionsGot, httpProtocolOptionsWantedAny)
```

nak3 commented on pkg/config/ext_authz_test.go:
```diff
@@ -0,0 +1,316 @@
+			}
+
+			if !reflect.DeepEqual(got.GetTypedConfig(), extAuthzWantedAny) {
+				t.Errorf("externalAuthZFilter() extAuthz = %v, extAuthzWantedAny %v", got.GetTypedConfig(), extAuthzWantedAny)
```
```suggestion
				t.Errorf("externalAuthZFilter() extAuthz = %v, want = %v", got.GetTypedConfig(), extAuthzWantedAny)
```

nak3 reviewed (commented):
Thank you so much @norbjd ! LGTM, I just left some comments for the style again.

nak3 commented:
> Also, I don't know if I should test https protocol, since it's the same as http.

I think current test cases without https are fine. Thank you!

nak3 commented:
/lgtm
/approve

Thank you @norbjd !
